conda-store-server/conda_store_server/build.py

```diff
@@ -110,6 +111,108 @@ def build_environment(conda_command, environment_filename, conda_prefix):
     )
 
 
+def build_lock_environment(lock_filename, conda_prefix):
+    return subprocess.check_output(
+        ["conda-lock", "install", "--prefix", conda_prefix, lock_filename],
+        stderr=subprocess.STDOUT,
+        encoding="utf-8",
+    )
+
+
+def fetch_and_extract_packages(conda_store, lock_filename):
+    """Fetch links from a conda-locked build recipe and then
+    gets a filelock on the required folder.
+    """
+    import libarchive
+    import tarfile
+    import shutil
+    from conda_store_server.conda import conda_root_package_dir
+
+    extensions = {".tar": "r:", ".tar.gz": "r:gz", ".tar.bz2": "r:bz2"}
+
+    prefix = conda_root_package_dir()
+
+    try:
+        prefix.exists()
+    except FileNotFoundError as e:
+        conda_store.log.error(
+            f"The conda prefix {prefix} does not exist. Traceback: {e}"
+        )
+
+    spec = dict()
+
+    with open(pathlib.Path(lock_filename)) as f:
+        spec = yaml.safe_load(f)
+
+    packages_searched = 1
+
+    for p in spec["package"]:
+        if p["manager"] != "conda":
+            # ignore non-conda managed packages for now
+            pass
+
+        else:
+            url = p["url"]
+
+            filename = pathlib.Path(url.split("/")[-1:][0])
+            filepath = prefix.joinpath(filename)
+            lockfile = filelock.FileLock(f"{filepath.__str__()}.lock")
+
+            with lockfile:
+
+                if filepath.exists():
+                    conda_store.log.info(f"SKIPPING {filename} | FILE EXISTS")
+                    packages_searched += 1
+
+                else:
+                    conda_store.log.info(
+                        f"DOWNLOAD {filename} | {packages_searched} of {len(spec['package'])}"
+                    )
+                    res = requests.get(url, stream=True)
+
+                    with open(filepath, "wb") as tarball:
+                        shutil.copyfileobj(res.raw, tarball)
+                        packages_searched += 1
+
+                    if ".tar" in filepath.__str__():
+                        filesuff = filename.suffixes
+                        file_extension = "".join(filesuff[filesuff.index(".tar") :])
+
+                        folderpath = pathlib.Path(
+                            str(filepath).replace(file_extension, "")
+                        )
+                        folderpath.mkdir(parents=True, exist_ok=False)
+
+                        conda_store.log.info(f"EXTRACT {filename} | PATH: {folderpath}")
+                        with tarfile.open(filepath, extensions[file_extension]) as tf:
+                            tf.extractall(folderpath)
+
+                    if ".conda" in filepath.__str__():
+                        libarchive.extract_file(filepath)
+                        for zip in filepath.glob('*.zst'):
+                            if "info" in str(zip):
```

Comment:
Could we use a more robust regex here? Or something more strict such as startswith/endswith etc.